ASAP.py

```diff
@@ -0,0 +1,312 @@
+#!/usr/bin/env python
+
+import math
+import numpy.fft
+from decimal import Decimal
+
+def smooth(data, resolution=1000):
+    ildr        = int( len(data) / resolution )
+    data        = SMA(data,ildr,ildr)
+    acf         = ACF(data, round(len(data)/10))
+    peaks       = acf.peaks
+    orig_kurt   = acf.kurtosis
+    min_obj     = acf.roughness
+    window_size = 1
+    lb          = 1
+
+    largest_feasible = -1
+    tail = len(data)/10
+    for i in range(len(peaks)-1,-1,-1):
+        w = peaks[i]
+
+        if w<lb or w == 1:
+            break
+
+        elif math.sqrt(1 - acf.correlations[w]) * window_size > math.sqrt(1-acf.correlations[window_size])*w:
+            continue
+
+        smoothed = SMA(data, w, 1)
+        metrics = Metrics(smoothed)
+        if metrics.roughness < min_obj and metrics.kurtosis >= orig_kurt:
+            min_obj = metrics.roughness
+            window_size = w
+            lb = round( max(w*math.sqrt( (acf.max_acf -1) / (acf.correlations[w]-1) ), lb) )
+
+    if largest_feasible > 0:
+        if largest_feasible < len(peaks)-2:
+            tail = peaks[largest_feasible +1]
+        lb = max(lb, peaks[largest_feasible]+1)
+
+    window_size = binary_search(lb, tail, data, min_obj, orig_kurt, window_size)
+    return SMA(data, window_size, 1)
+
+
+def binary_search(head,tail,data,min_obj,orig_kurt,window_size):
+    while head <= tail:
+        w = round((head+tail)/2.0)
+        smoothed = SMA(data,w,1)
+        metrics  = Metrics(smoothed)
+        if metrics.kurtosis >= orig_kurt:
+            if metrics.roughness < min_obj:
+                window_size = w
+                min_obj = metrics.roughness
+            head = w + 1
+        else:
+            tail = w - 1
+    return window_size
+
+def SMA(data, _range, slide):
```

Comment:
I'm ok with this for sure.

I hadn't put much thought into optimizations at all. I was just trying get it ported more-or-less as-is.

I'll update the PR today.